WordPress/src/main/java/org/wordpress/android/ui/posts/PostsPagerAdapter.kt

```diff
@@ -0,0 +1,38 @@
+package org.wordpress.android.ui.posts
+
+import android.content.Context
+import android.support.v4.app.FragmentManager
+import android.support.v4.app.FragmentPagerAdapter
+import android.view.ViewGroup
+import org.wordpress.android.fluxc.model.SiteModel
+import org.wordpress.android.ui.posts.PostListType.DRAFTS
+import org.wordpress.android.ui.posts.PostListType.PUBLISHED
+import org.wordpress.android.ui.posts.PostListType.SCHEDULED
+import org.wordpress.android.ui.posts.PostListType.TRASHED
+import java.lang.ref.WeakReference
+
+class PostsPagerAdapter(private val site: SiteModel, val context: Context, val fm: FragmentManager) :
+        FragmentPagerAdapter(fm) {
+    companion object {
+        val postTypes = listOf(PUBLISHED, DRAFTS, SCHEDULED, TRASHED)
+    }
+
+    private val listFragments = mutableMapOf<Int, WeakReference<PostListFragment>>()
+
+    override fun getCount(): Int = postTypes.size
+
+    override fun getItem(position: Int): PostListFragment =
+            PostListFragment.newInstance(site, postTypes[position], null)
+
+    override fun instantiateItem(container: ViewGroup, position: Int): Any {
+        val fragment = super.instantiateItem(container, position) as PostListFragment
+        listFragments[position] = WeakReference(fragment)
```

Comment:
Shouldn't this happen in the `getItem` instead? When I put a breakpoint here, it gets triggered almost every time I swipe between tabs, so it's definitely not called only once. I thought the fragments were getting destroyed, but `getItem` is getting called only once, so I am not sure what's going on. If you have an insight I'd really appreciate it, as otherwise I am going to dig into this a bit.

P.S: The naming in `FragmentPagerAdapter` is super confusing to me. Just from the names, I feel `getItem` should just return the already created fragments and `instantiateItem` is the one that should be creating the fragments with `newInstance` however the code doesn't behave that way as far as I can tell. The docs are not very helpful either.

Reply:
I completely agree with your "P.S." :D.. The naming is extremely confusing.

- `ViewPager` caches up to 2 fragments by default (1 fragment on the left of the current fragment and 1 fragment on the right of the current fragment)

FragmentPagerAdapter
- `instantiateItem` asks the `FragmentManager` to find the fragment -> `FragmentManager` searches fragments attached to the current Activity and if it's not there it searches all the fragments on the BackStack (Sidenote: fragments on the backstack might not be alive anymore). -> if the `FragmentManager` doesn't find the fragment, the `getItem` method is called.

Examples:) - Imagine we enter the screen and fragment A is displayed 
`instantiateItem` for A and B is called (`ViewPager` wants to preload B it into the cache). Since this is the first time we load both A and B, `getItem` is called for both.

**A** B C -> A **B** C
`instantiateItem` for C is called (`ViewPager` wants to preload it into the cache). Since this is the first time we load C, `getItem` is called.

A **B** C -> **A** B C 
Neither `instantiateItem` nor `getItem` is called since A is cached in `ViewPager`

**A** B C -> A **B** C
`instantiateItem` for C is called, but `getItem` is not called since the C fragment is found in the `FragmentManager`


>Shouldn't this happen in the `getItem` instead
That's what I thought at first, but the problem is rotation :(. When we rotate the device, we lose content of `listFragments` variable and `getItem` is not called since the fragment is found in the `FragmentManager`. Having `listFragments[position] = WeakReference(fragment)` in the `instantiateItem` is definitely a waste of a couple of references, on the other hand since this action is tied to the user interaction (it won't happen million times) I thought it doesn't really matter. Wdyt?

Reply:
Thank you for this amazing explanation 🙇 I agree, this seems to be the most reasonable solution. Let's go with this and if we find out anything else about this (any issues during testing) we can re-visit it.